horovod/common/operations.cc

```diff
@@ -1545,15 +1396,16 @@ bool RunLoopOnce(HorovodGlobalState& state, MPIContext& ctx,
     // ranks at this tick.
 
     // 1. Get message lengths from every rank.
```

Comment:
I think having the Controller be an abstraction over collective ops like Reduce and Gather is too granular.  Ideally, the controller should encapsulate the logic of Parts 1, 2, 3, and 5 here are logically doing one thing: collecting messages.  So my suggestion would be to put all of this behind a single method in MPIController called `CollectMessage()` or similar.  What do you think?